emu_sv/time_evolution.py

```diff
@@ -87,22 +87,19 @@ def __matmul__(self, vec: torch.Tensor) -> torch.Tensor:
 class DHDDeltaSparse:
     """
     Derivative of the Rydberg Hamiltonian respect to Delta:
-        ∂H/∂Δₖ = -nₖ
+        ∂H/∂Δᵢ = -nᵢ
     """
 
-    def __init__(self, index: int, device: torch.device, nqubits: int):
-        self.index = index
-        self.shape = (2**index, 2, 2 ** (nqubits - index - 1))
-        diag = torch.zeros(
-            *self.shape,
-            dtype=torch.complex128,
-            device=device,
-        )
-        diag[:, 1] = -1.0
-        self.diag = diag.reshape(-1)
+    def __init__(self, i: int, nqubits: int):
+        self.nqubits = nqubits
+        self.shape = (2**i, 2, 2 ** (nqubits - i - 1))
 
     def __matmul__(self, vec: torch.Tensor) -> torch.Tensor:
-        return vec * self.diag
+        result = vec.clone()
+        result = result.reshape(vec.shape[0], *self.shape)
+        result[:, :, 0] = 0.0
+        result[:, :, 1] *= -1.0
```

Comment:
can't you just multiply the whole of `result`? since 0 * x == 0
`result *= -1.0`
not completely sure what is better but I think avoiding the slicing might actually be more efficient